MediaGalleryUi/Controller/Adminhtml/Directories/GetTree.php

```diff
@@ -0,0 +1,163 @@
+    public function __construct(
+        Action\Context $context,
+        LoggerInterface $logger,
+        Filesystem $filesystem,
+        string $path
+    ) {
+        parent::__construct($context);
+        $this->logger = $logger;
+        $this->filesystem = $filesystem;
+        $this->path = $path;
+        $this->responseContent = [];
+    }
+    /**
+     * @inheritdoc
+     */
+    public function execute()
+    {
+        try {
+            $directories = [];
+            $directoryInstance = $this->filesystem->getDirectoryRead($this->path);
+            if ($directoryInstance->isDirectory()) {
+                foreach ($directoryInstance->readRecursively() as $index => $path) {
+                    if ($directoryInstance->isDirectory($path)) {
+                        $pathArray = explode('/', $path);
+                        $directories[] =
+                            [
+                                'data' => count($pathArray) > 0 ? end($pathArray) : $path,
+                                'attr' => ['id' => $index],
+                                'metadata' => [
+                                    'path' => $path
+                                ],
+                                'path_array' => $pathArray
+                            ];
+                    }
+                }
+            }
+            $this->responseContent = $this->buildFolderTree($directories);
+
+            $responseCode = self::HTTP_OK;
+        } catch (\Exception $exception) {
+            $this->logger->critical($exception);
+            $responseCode = self::HTTP_INTERNAL_ERROR;
+            $this->responseContent = [
+                'success' => false,
+                'message' => __('Retrieving directories list failed.'),
+            ];
+        }
+
+        /** @var Json $resultJson */
+        $resultJson = $this->resultFactory->create(ResultFactory::TYPE_JSON);
+        $resultJson->setHttpResponseCode($responseCode);
+        $resultJson->setData($this->responseContent);
+
+        return $resultJson;
+    }
+
+    /**
+     * Build folder tree structure by provided directories path
+     *
+     * @param array $directories
+     * @return array
+     */
+    private function buildFolderTree(array $directories, bool $skipRoot = true): array
+    {
+        $tree = [
+            'name' => 'root',
+            'path' => '/',
+            'children' => []
+        ];
+        foreach ($directories as $idx => &$node) {
+            $node['children'] = [];
+            $result = $this->findParent($node, $tree);
+            $parent = & $result['treeNode'];
+
+            $parent['children'][] =& $directories[$idx];
+        }
+        return $skipRoot ? $tree['children'] : $tree;
+    }
+
+    /**
+     * Find parent directory
+     *
+     * @param array $node
+     * @param array $treeNode
+     * @param int $level
+     * @return array
+     */
+    private function findParent(array &$node, array &$treeNode, int $level = 0): array
```

Comment:
I believe the custom implementation of recursive directory iteration would help to reduce the algorithm complexity (no need to search for parent if the algorithm builds tree configuration level by level).